osbenchmark/builder/installers/opensearch_installer.py

```diff
@@ -26,9 +26,10 @@ def __init__(self, provision_config_instance, executor, hook_handler_class=Boots
         self.config_applier = ConfigApplier(executor, self.template_renderer, self.path_manager)
         self.host_cleaner = HostCleaner(self.path_manager)
 
-    def install(self, host, binaries, all_node_ips):
+    # pylint: disable=arguments-differ
+    def install(self, host, binaries, all_node_ips, config_vars=None):
```

Comment:
Should `config_vars` just be added to the `Installer.install` interface? 

Is there a reason that disabling the linter is preferred?

Reply:
`config_vars` shouldn't be part of this method's signature at all IMO. An installer should be responsible for owning and defining its own config_vars rather than having some passed into it.

The trouble comes when both OS and plugin(s) are set to be installed. In this case, the config_vars [from all sources must be considered](https://github.com/opensearch-project/opensearch-benchmark/blob/main/osbenchmark/builder/provisioner.py#L215-L240). Adding `config_vars` to the signature allows the CompositeInstaller to pass in the finalized config_vars from all sources without changing the general flow of the sub-installers. OpenSearchInstaller can be consumed as a standalone Installer implementation or utilized by the CompositeInstaller via the same install method.

I feel that I am missing a design pattern that would eliminate the need to add `config_vars` to this method. I will think on it some more and try to come up with a better solution.

Reply:
I agree with your assessment that `config_vars` shouldn't be a part of this method. Its essentially an implementation detail that the user of this method shouldn't care about. Also, that's why the `Installer` interface shouldn't & doesn't have it.

Please correct me if I'm wrong, but it seems like each installer already just makes a single install call in its lifetime. Constructor is instance specific and can be different for classes implementing the same interface, have you considered having the `config_vars` as an optional constructor param for `OpenSearchInstaller` and making it an class level variable? 

I believe this way the CompositeInstaller can have multiple sub-installers, and not break their general flow. However, if you mean the CompositeInstaller needs results from sub-installers to come up with the final `config_vars`, I'd still recommend the same approach, but using a factory class the generates the OS installer instance after the `config_vars` are ready.

Reply:
The CompositeInstaller does need results from the sub-installers to come up with the final `config_vars`. Specifically, it needs the Node object generated by the OpenSearchInstaller's `install` method.

With the current implementation there's no way to generate the `config_vars` without first instantiating an instance of OpenSearchInstaller and invoking the `install` method to get a Node. I think a further layer of abstraction to separate preparing the installation and applying the config data is the solution here: create new OpenSearchPreparer and PluginPreparer classes with `prepare` and `get_config_vars` methods. 

Then the OpenSearch and PluginInstallers would take in a respective instance of these Preparer classes and perform the whole installation flow (including applying the config) within the install method. The CompositeInstaller would have instances of both OpenSearchPreparer and PluginPreparer, with the flow being: call prepare methods -> call `get_config_vars` methods -> generate final `config_vars` -> apply configs

I plan to make heavy use of the factory pattern when selecting the correct builder components for the `ClusterBuilder` :)